### PR TITLE
[R4R] Fix memory consume during taking snapshot for state sync

### PR DESCRIPTION
**Pull request by ackratos**
discard any reference to the Node generated during traversing iavl tree because state sync only needs serialized node. But we have to decode beforehand to get left and right hash.

This reduces memory cost for snapshotting height 451000 in QA where has 10000000 iavl nodes (5MM accounts) from more than 2G to less than 50M 

before: 
profile001

after:
profile002

### CODE REVIEW

File: node.go [hunk 1]
```diff
@@ -382,8 +382,8 @@ func (node *Node) traverse(t *ImmutableTree, ascending bool, cb func(*Node) bool
 }
 
 // traverseFirst is a wrapper over traverseInRange when we want the whole tree and will traverse the leaf nodes
-func (node *Node) traverseFirst(t *ImmutableTree, ascending bool, cb func(*Node) bool) bool {
-	return node.traverseInRange(t, nil, nil, ascending, false, 0, func(node *Node, depth uint8) bool {
+func (node *Node) traverseForStateSync(t *ImmutableTree, ascending bool, cb func(*Node) bool) bool {
```
Comment (rickyyangz) [inline node.go]: what is `statesync` for this iavl lib ?

Comment (ackratos) [reply node.go]: > what is `statesync` for this iavl lib ?

reverted

File: node.go [hunk 2]
```diff
@@ -439,6 +439,85 @@ func (node *Node) traverseInRange(t *ImmutableTree, start, end []byte, ascending
 	return stop
 }
 
+// This method doesn't hold any reference to loaded node, its cb's responsibility to decide whether hold Node reference
+// If cb (like usage in state sync) don't need hold a reference to loaded Node, the memory footprint will be near to zero
+func (node *Node) traverseInRangeDiscardNode(t *ImmutableTree, start, end []byte, ascending bool, inclusive bool, depth uint8, cb func(*Node, uint8) bool) bool {
```
Comment (guagualvcha) [inline node.go]: Total agree with your logic.  What if we can reuse the origin code :
```
func (node *Node) traverseInRangeDiscardNode(t *ImmutableTree, start, end []byte, ascending bool, inclusive bool, depth uint8, cb func(*Node, uint8) bool) bool {
defer func(){
		node.leftNode = nil
		node.rightNode = nil
}()
     return node.traverseInRange(t, nil, nil, ascending, false, 0, func(node *Node, depth uint8) bool {
		return cb(node)
	})
}
```

**Comment by ackratos** [comment]
fix missing hash panic

fix

File: node.go (hunk 2 above)
Comment (rickyyangz) [reply node.go]: not cheap to use defer, especially when there is a recursive call

**Pull request merged**